flashcards/textualui.py

```diff
@@ -0,0 +1,233 @@
+        )
+
+
+class InfoPane(Widget):
+    """
+    Widget to show one line of text information
+    """
+
+    default_style_text = "bright_white on bright_blue"
+    blink_style_text = f"blink {default_style_text}"
+
+    def __init__(self):
+        super().__init__()
+        self.text = ""
+        self.style_text = InfoPane.default_style_text
+
+    def blink(self, enabled):
+        """
+        Turn blink on or off for the text in this pane
+        """
+        self.style_text = (
+            InfoPane.blink_style_text if enabled else InfoPane.default_style_text
+        )
+        self.refresh()
+
+    def render(self) -> RenderableType:
+        return Padding(
+            pad=(1, 0, 1, 0),
+            renderable=Text(
+                text=self.text,
+                justify="center",
+            ),
+            # colors: https://github.com/Textualize/rich/blob/master/docs/source/appendix/colors.rst
+            style=self.style_text,
+        )
+
+
+class StatusBar(Widget):
+    """
+    Status bar widget
+    """
+
+    def __init__(self):
+        super().__init__()
+        self.text = ""
+
+    def render(self) -> RenderableType:
+        return Padding(
+            renderable=Text(
+                text=self.text,
+                justify="right",
+            ),
+            pad=(0, 1, 0, 0),
+            # colors: https://github.com/Textualize/rich/blob/master/docs/source/appendix/colors.rst
+            style="bright_white on bright_black",
+        )
+
+
+class TextualUi(Ui, App):
+    """
+    Interface to interact with the user in the flashcard game, in a console,
+    using the textual library
+    """
+
+    def __init__(self, translator: Translator):
+        Ui.__init__(self)
+        App.__init__(self, True, None, log="foo.txt")
+        self._ = translator
+        self.flashcard_text = ""
+        self.text_input = TextInput(
+            name="",
+            placeholder=self._("guess_prompt"),
+            title="",
+        )
+        self.flashcard = FlashcardWidget()
+        self.info_pane1 = InfoPane()
+        self.info_pane2 = InfoPane()
+        self.status_bar = StatusBar()
+
+    async def on_mount(self) -> None:
+        """
+        Layout the UI
+        """
+        await self.view.dock(self.status_bar, edge="bottom", size=1)
+        await self.view.dock(self.info_pane2, edge="bottom", size=5)
+        await self.view.dock(self.text_input, edge="bottom", size=3)
+        await self.view.dock(self.info_pane1, edge="top", size=5)
+        await self.view.dock(self.flashcard, edge="top")
+        await self.text_input.focus()
+
+    def setup(self, max_key_length: int, max_answer_length: int):
+        self.flashcard.max_text_size = max_key_length
+
+    def new_game(self):
+        self.info_pane1.text = ""
+        self.info_pane2.text = ""
+        self.info_pane2.blink(False)
+        self.text_input.value = ""
+        self.text_input.visible = True
```

Comment:
This isn't working: after the end of the game, if the user chooses "y" to continue quizzing the missed cards, the input remains hidden. Resizing the terminal fixes it...